collect_app/src/main/java/org/odk/collect/android/google/GoogleAccountsManager.java

```diff
@@ -0,0 +1,249 @@
+
+    private HttpTransport transport;
+    private JsonFactory jsonFactory;
+
+    private boolean autoChooseAccount = true;
+    private GeneralSharedPreferences preferences;
+    private Intent intentChooseAccount;
+
+    public GoogleAccountsManager(@NonNull Activity activity) {
+        this.activity = activity;
+        initCredential(activity);
+    }
+
+    public GoogleAccountsManager(@NonNull Fragment fragment) {
+        this.fragment = fragment;
+        initCredential(fragment.getActivity());
+    }
+
+    public GoogleAccountsManager(@NonNull Context context) {
+        initCredential(context);
+    }
+
+    /**
+     * This constructor should be used only for testing purposes
+     */
+    public GoogleAccountsManager(@NonNull GoogleAccountCredential credential,
+                                 @NonNull GeneralSharedPreferences preferences,
+                                 @NonNull Intent intentChooseAccount) {
+        this.credential = credential;
+        this.preferences = preferences;
+        this.intentChooseAccount = intentChooseAccount;
+    }
+
+    private void initCredential(Context context) {
+        this.context = context;
+
+        transport = AndroidHttp.newCompatibleTransport();
+        jsonFactory = JacksonFactory.getDefaultInstance();
+        preferences = GeneralSharedPreferences.getInstance();
+
+        credential = GoogleAccountCredential
+                .usingOAuth2(context, Collections.singletonList(DriveScopes.DRIVE))
+                .setBackOff(new ExponentialBackOff());
+
+        intentChooseAccount = credential.newChooseAccountIntent();
+    }
+
+    public void setSelectedAccountName(String accountName) {
+        if (accountName != null) {
+            preferences.save(PreferenceKeys.KEY_SELECTED_GOOGLE_ACCOUNT, accountName);
+            selectAccount(accountName);
+        }
+    }
+
+    @Override
+    public void onPermissionsGranted(int requestCode, List<String> list) {
+        if (listener != null) {
+            listener.onGoogleAccountSelected(credential.getSelectedAccountName());
+        }
+    }
+
+    @Override
+    public void onPermissionsDenied(int requestCode, List<String> list) {
+        ToastUtils.showShortToast("Permission denied");
+    }
+
+    public void onRequestPermissionsResult(int requestCode,
+                                           @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        EasyPermissions.onRequestPermissionsResult(
+                requestCode, permissions, grantResults, this);
+    }
+
+    public void chooseAccount() {
+        if (hasPermissions()) {
+            String accountName = getSelectedAccount();
+            if (autoChooseAccount && !accountName.isEmpty()) {
+                selectAccount(accountName);
+            } else {
+                showAccountPickerDialog();
+            }
+        }
+    }
+
+    public boolean hasPermissions() {
```

Comment:
Its not clear from this method's name that it has a side effect (that is checks for permissions). I would swap 'hasPermissions' which implies an accessor and 'checkAccountPermissions' which implies an action, or consider other names.